text/3541-async-drop.md

```diff
@@ -0,0 +1,222 @@
+
+```rust
+use async_trait::async_trait;
+use tokio::time::{sleep, Duration};
+
+// This trait should be implement internally
+#[async_trait]
+trait AsyncDrop {
+    async fn drop(&mut self);
+}
+
+// Define an AsyncResource struct implementing AsyncDrop
+struct AsyncResource {
+    data: String,
+}
+
+// Implement AsyncDrop for AsyncResource
+#[async_trait]
+impl AsyncDrop for AsyncResource {
+    async fn drop(&mut self) {
+        println!("Async cleanup for AsyncResource: {}", self.data);
+        // Asynchronous cleanup logic, e.g., releasing resources
+    }
+}
+
+#[tokio::main]
+async fn main() {
+    {
+        let _async_res = AsyncResource {
+            data: String::from("Example Data"),
+        };
+        // drop will be triggered here
+    }
+    // AsyncResource is dropped as it goes out of scope, and its asynchronous drop logic may be executed later by the Tokio runtime
+
+    // Sleep to ensure `drop` completed its execution
+    sleep(Duration::from_secs(2)).await;
+    println!("Main function completed");
+}
+```
+
+Moreover, to seamlessly integrate native async drop support into Rust, it is essential to outline how executors should handle asynchronous resource cleanup. This section also provides guidelines for developers using executors to ensure consistent and correct execution of async drop logic.
+
+1. **Recognition of AsyncDrop Trait**
+   Executors, such as `MyExecutor` in the following example, should be updated to recognize types implementing the `AsyncDrop` trait. This can be achieved through a mechanism such as trait bounds or associated types to identify futures that require special handling for asynchronous cleanup.
+
+   ```rust
+   trait Executor {
+       fn spawn(&self, future: impl Future);
+   }
+
+   impl<T: AsyncDrop> Executor for MyExecutor {
+       // Your custom cleanup logic goes here
+   }
+   ```
+
+   Here, `T: AsyncDrop` indicates that this executor can work with futures (`T`) that implement the AsyncDrop trait. The `AsyncDrop` trait serves as a marker, signaling that the associated type requires special handling for asynchronous cleanup.
+
+1. **Async Drop Invocation**
+   Executors must ensure that the async `drop` method is invoked when a future implementing the `AsyncDrop` trait goes out of scope or completes. This involves tracking the lifecycle of futures and, upon termination, triggering the async drop logic before releasing associated resources.
+
+   ```rust
+   async fn execute_future<T: AsyncDrop>(&self, fut: T) {
+       // Execute the future
+       let result = block_on(fut);
+   
+       // Perform async drop before releasing resources
+       fut.drop().await;
+   }
+   ```
```

Comment:
This code cannot be valid. Ownership of `fut` was passed to `block_on`, so you can't call methods like `.drop()` on it afterwards.

Moreover `block_on` is a terrible practice inside an `async fn`. Why use it when you can just `.await`? If instead you can't `.await` `fut`, then you also can't `.await` `fut.drop()`.